pkg/pool-weighted/contracts/smart/InvestmentPool.sol

```diff
@@ -0,0 +1,259 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/ReentrancyGuard.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/helpers/WordCodec.sol";
+
+import "../BaseWeightedPool.sol";
+import "./WeightCompression.sol";
+
+/**
+ * @dev Weighted Pool with mutable weights, designed to support investment use cases: large token counts,
+ * rebalancing through gradual weight updates, and enabling/disabling trading.
+ */
+contract InvestmentPool is BaseWeightedPool, ReentrancyGuard {
+    // The Pause Window and Buffer Period are timestamp-based: they should not be relied upon for sub-minute accuracy.
+    // solhint-disable not-rely-on-time
+
+    using FixedPoint for uint256;
+    using WordCodec for bytes32;
+    using WeightCompression for uint256;
+
+    // State variables
+
+    // Cached here to avoid calling getTokens on the pool, which would be very gas-intensive for large numbers of tokens
+    // Can only change if tokens are added/removed
+    uint256 private _tokenCountCache;
+
+    // Store scaling factor and start/end weights for each token
+    // Mapping should be more efficient than trying to compress it further
+    // into a fixed array of bytes32 or something like that, especially
+    // since tokens can be added/removed - and re-ordered in the process
+    // For each token, we store:
+    // [ 155 bits|   5 bits |  32 bits   |   64 bits    |
+    // [ unused  | decimals | end weight | start weight |
+    // |MSB                                          LSB|
+    mapping(IERC20 => bytes32) private _poolState;
+
+    uint256 private constant _START_WEIGHT_OFFSET = 0;
+    uint256 private constant _END_WEIGHT_OFFSET = 64;
+    uint256 private constant _DECIMAL_DIFF_OFFSET = 96;
+
+    // Time travel comment
+    // [ 192 bits | 32 bits  |  32 bits   |
+    // [  unused  | end time | start time |
+    // |MSB                            LSB|
+    //bytes32 private _gradualUpdateTimestamps;
+
+    //uint256 private constant _START_TIME_OFFSET = 0;
+    //uint256 private constant _END_TIME_OFFSET = 32;
```

Comment:
Is this from another branch?

Reply:
Commented out stuff that says time travel anticipates the next PR (included for context, since otherwise it would be really weird to call functions like startGradualWeightUpdate to initialize weights, when the weight update functions aren't there yet